### ISSUE TITLE
StepFunctions: Unable to parse and update resource

### ISSUE THREAD

**Issue by eugene-nn** [localstack/localstack#3249]
Hello!

I'm currently experiencing an issue with step functions update config behaviour on localstack. 

my docker-compose file definition:

```version

services:
  localstack:
    container_name: localstack_main
    image: localstack/localstack
    network_mode: bridge
    ports:
      - "4566:4566"
    environment:
      - DEBUG=1
      - DOCKER_HOST=unix:///var/run/docker.sock
    volumes:
      - "${TMPDIR:-/tmp/localstack}:/tmp/localstack"
      - "/var/run/docker.sock:/var/run/docker.sock"
```

serverless.yaml:

```service

plugins:
  - serverless-localstack
  - serverless-step-functions

custom:
  localstack:
    stages:
      - local
    host: http://localhost
    edgePort: 4566

provider:
  name: aws
  profile: default
  runtime: nodejs12.x
  stage: local
  region: us-east-1

stepFunctions:
  stateMachines:
    flow:
      definition:
        Comment: 'Simple flow'
        StartAt: Start
        States:
          Start:
            Type: Pass
            Next: Stop
#          Passthru:
#            Type: Pass
#            Next: Stop
          Stop:
            Type: Pass
            End: true
```

As you can see I have simple state machine consisting of two steps(just for test) and no error was registered on localstack yet.

Output of `sls deploy --stage local -v` looks like:

```Serverless
Serverless: Warning: Unable to find plugin named: TypeScriptPlugin
Serverless: Packaging service...
Serverless: Creating Stack...
Serverless: Checking Stack create progress...
CloudFormation - CREATE_COMPLETE - AWS::CloudFormation::Stack - flow-local
Serverless: Stack create finished...
Serverless: Uploading CloudFormation file to S3...
Serverless: Uploading artifacts...
Serverless: Validating template...
Serverless: Skipping template validation: Unsupported in Localstack
Serverless: Updating Stack...
Serverless: Checking Stack update progress...
CloudFormation - UPDATE_COMPLETE - AWS::CloudFormation::Stack - flow-local
Serverless: Stack update finished...
Service Information
service: flow
stage: local
region: us-east-1
stack: flow-local
resources: 4
api keys:
  None
endpoints:
  None
functions:
  None
layers:
  None

Stack Outputs
ServerlessDeploymentBucketName: flow-local-serverlessdeploymentbucket-n67i63454co7
FlowStepFunctionsStateMachineArn: arn:aws:states:us-east-1:000000000000:stateMachine:flow-local-FlowStepFunctionsStateMachine-LXOPB7601I5A
```

Then after modifying existing state machine (removing comments on Passthru section):

```
...
stepFunctions:
  stateMachines:
    flow:
      definition:
        Comment: 'Simple flow'
        StartAt: Start
        States:
          Start:
            Type: Pass
            Next: Passthru
          Passthru:
            Type: Pass
            Next: Stop
          Stop:
            Type: Pass
            End: true
```

And after redeploy it I can see error:

```
2020-11-16T20:16:21:ERROR:localstack.services.cloudformation.cloudformation_starter: Unable to parse and update resource "FlowStepFunctionsStateMachine": type object 'StateMachine' has no attribute 'update_from_cloudformation_json' Traceback (most recent call last):
  File "/opt/code/localstack/localstack/services/cloudformation/cloudformation_starter.py", line 295, in parse_and_update_resource
    return _parse_and_create_resource(logical_id, resource_json, resources_map, region_name, update=True)
  File "/opt/code/localstack/localstack/services/cloudformation/cloudformation_starter.py", line 337, in _parse_and_create_resource
    parse_and_update_resource_orig(logical_id, resource_json_arns_fixed, resources_map, region_name)
  File "/opt/code/localstack/.venv/lib/python3.8/site-packages/moto/cloudformation/parsing.py", line 338, in parse_and_update_resource
    resource_class.update_from_cloudformation_json, "__isabstractmethod__"
AttributeError: type object 'StateMachine' has no attribute 'update_from_cloudformation_json'
```

And just for double check if I trigger command: 

```
aws --endpoint-url=http://localhost:4566 stepfunctions describe-state-machine --state-machine-arn $(aws --endpoint-url=http://localhost:4566 stepfunctions list-state-machines | jq -r '.stateMachines[].stateMachineArn') | jq .definition
```

I will see the initial state machine definition:

```
"{\n  \"Comment\": \"Simple flow\",\n  \"StartAt\": \"Start\",\n  \"States\": {\n    \"Start\": {\n      \"Type\": \"Pass\",\n      \"Next\": \"Stop\"\n    },\n    \"Stop\": {\n      \"Type\": \"Pass\",\n      \"End\": true\n    }\n  }\n}"
```

prettified version:

```
{
  Comment: 'Simple flow',
  StartAt: 'Start',
  States: {
    Start: { Type: 'Pass', Next: 'Stop' },
    Stop: { Type: 'Pass', End: true }
  }
}
```

which is not containing latest added step - `Passthru`. 

Can you give advice on debugging the issue? Thanks

┆Issue is synchronized with this [Jira Bug](https://localstack.atlassian.net/browse/LOC-326) by [Unito](https://www.unito.io/learn-more)

**Comment by whummer** [comment]
Thanks for reporting @eugene-nn . This should be fixed in #3341 - can you please pull the `latest` Docker image and give it another try? Please report here if the problem persists. Thanks

**Comment by eugene-nn** [comment]
Hi @whummer @duongpv7 
Seems it works BUT only for step functions which state's type is Pass.  Now I can new add state and updated stack will have my changes. If case of type is Task(you can check example below) it will not work anymore. However it was before. 

```...
functions:
  fn:
    handler: src/handler/index.handler
...
stepFunctions:
  stateMachines:
    flow:
      definition:
        Comment: 'Simple flow'
        StartAt: Start
        States:
          Start:
            Type: Task
            Resource:
              Fn::GetAtt: [fn, Arn]
            Next: Passthru
          Passthru:
            Type: Pass
            Next: Stop
          Stop:
            Type: Pass
            End: true
```
Error:
```
localstack_main | 2020-12-13T14:37:42:ERROR:localstack.services.cloudformation.cloudformation_starter: Unable to parse and create resource "FlowStepFunctionsStateMachine": replace() argument 2 must be str, not dict Traceback (most recent call last):
localstack_main |   File "/opt/code/localstack/localstack/services/cloudformation/cloudformation_starter.py", line 374, in parse_and_create_resource
localstack_main |     result = _parse_and_create_resource(
localstack_main |   File "/opt/code/localstack/localstack/services/cloudformation/cloudformation_starter.py", line 438, in _parse_and_create_resource
localstack_main |     resource_tuple = parsing.parse_resource_and_generate_name(logical_id, resource_json, resources_map)
localstack_main |   File "/opt/code/localstack/.venv/lib/python3.8/site-packages/moto/cloudformation/parsing.py", line 282, in parse_resource_and_generate_name
localstack_main |     resource_tuple = parse_resource(resource_json, resources_map)
localstack_main |   File "/opt/code/localstack/.venv/lib/python3.8/site-packages/moto/cloudformation/parsing.py", line 274, in parse_resource
localstack_main |     resource_json = clean_json(resource_json, resources_map)
localstack_main |   File "/opt/code/localstack/localstack/services/cloudformation/cloudformation_starter.py", line 299, in clean_json
localstack_main |     rs = clean_json_orig(resource_json, resources_map)
localstack_main |   File "/opt/code/localstack/.venv/lib/python3.8/site-packages/moto/cloudformation/parsing.py", line 202, in clean_json
localstack_main |     cleaned_val = clean_json(value, resources_map)
localstack_main |   File "/opt/code/localstack/localstack/services/cloudformation/cloudformation_starter.py", line 299, in clean_json
localstack_main |     rs = clean_json_orig(resource_json, resources_map)
localstack_main |   File "/opt/code/localstack/.venv/lib/python3.8/site-packages/moto/cloudformation/parsing.py", line 202, in clean_json
localstack_main |     cleaned_val = clean_json(value, resources_map)
localstack_main |   File "/opt/code/localstack/localstack/services/cloudformation/cloudformation_starter.py", line 349, in clean_json
localstack_main |     resource_json['Fn::Sub'][0] = resource_json['Fn::Sub'][0].replace('${%s}' % k, v)
localstack_main | TypeError: replace() argument 2 must be str, not dict
localstack_main |
localstack_main | 2020-12-13T14:37:42:INFO:localstack.services.cloudformation.cloudformation_starter: Error running stack deployment loop: replace() argument 2 must be str, not dict
localstack_main | 2020-12-13T14:37:42:WARNING:bootstrap.py: Thread run method <function apply_patches.<locals>.run_dependencies_deployment_loop.<locals>.run_loop at 0x7f62f4bdc1f0>(None) failed: Unable to resolve all CloudFormation resources after traversing dependency tree (maximum depth 40): [] Traceback (most recent call last):
localstack_main |   File "/opt/code/localstack/localstack/utils/bootstrap.py", line 581, in run
localstack_main |     result = self.func(self.params)
localstack_main |   File "/opt/code/localstack/localstack/services/cloudformation/cloudformation_starter.py", line 1210, in run_loop
localstack_main |     raise Exception('Unable to resolve all CloudFormation resources after traversing ' +
localstack_main | Exception: Unable to resolve all CloudFormation resources after traversing dependency tree (maximum depth 40): []
localstack_main |
```

**Comment by whummer** [comment]
Thanks for the update @eugene-nn . Another fix has been pushed in #3355 - can you please pull the `latest` Docker image and give it another try? Please let us know if the problem persists. Thanks for your help.

**Comment by eugene-nn** [comment]
Hi @whummer @duongpv7
Thank you guys, create/update scenarios are valid.
